### PR TITLE
#25 - Improved the Thread model

### PR DESCRIPTION
**Pull request by ddolcimascolo**

### CODE REVIEW

File: lib/models/Thread.js [hunk 1]
```diff
@@ -47,6 +47,88 @@ export default class Thread extends Model {
...
+      options[id] = updates;
+    });
+
+    return this._jmap.setMessages({ update: options });
+  }
+
+  _flagAllMessages(flag, state) {
+    Utils.assertRequiredParameterHasType(state, flag, 'boolean');
+
+    return this._updateAllMessages({ [flag]: state });
```
Comment (dbenchi) [inline lib/models/Thread.js:128]: Why do you need an Array of flag here?

File: lib/models/Thread.js [hunk 2]
```diff
@@ -47,6 +47,88 @@ export default class Thread extends Model {
...
+   *
+   * @return {Promise} A {@link Promise}, as per {@link Client.destroyMessages}.
+   *
+   * @see Client#destroyMessages
+   */
+  destroy() {
+    return this._jmap.destroyMessages(this.messageIds);
+  }
+
+  _updateAllMessages(updates) {
```
Comment (dbenchi) [inline lib/models/Thread.js:115]: The variable name updates is not well chosen!
By updates, you mean that you can change several message flags at once?

File: lib/models/Thread.js [hunk 3]
```diff
@@ -47,6 +47,88 @@ export default class Thread extends Model {
...
+
+    this.messageIds.forEach(id => {
+      options[id] = updates;
+    });
+
+    return this._jmap.setMessages({ update: options });
+  }
+
+  _flagAllMessages(flag, state) {
+    Utils.assertRequiredParameterHasType(state, flag, 'boolean');
```
Comment (dbenchi) [inline lib/models/Thread.js]: IT should be: Utils.assertRequiredParameterHasType('state', state, 'boolean');

Comment (dbenchi) [inline lib/models/Thread.js]: What about Utils.assertRequiredParameterHasType('flag', flag, 'string');

File: test/common/Client.js [hunk 4]
```diff
@@ -2022,4 +2022,59 @@ describe('The Client class', function() {
...
+          destroyed: ['id']
+        }, '#0']]);
+      };
+
+      client.destroyMessages(['id']).then(function(messagesSet) {
+        expect(messagesSet).to.be.an.instanceof(jmap.MessagesSet);
+
+        done();
+      });
+    });
```
Comment (dbenchi) [inline test/common/Client.js:2076]: 2 more tests to be added:
1- messages not destroyed with the reason in notDestroyed
2- messages not destroyed and the reason is messing

File: lib/models/Thread.js [hunk 5]
```diff
@@ -47,6 +47,88 @@ export default class Thread extends Model {
...
+   * Moves this {@link Thread} to a different set of mailboxes.
+   *
+   * @param mailboxIds {String[]} The identifiers of the target mailboxes for the thread.
+   *
+   * @return {Promise} A {@link Promise}, as per {@link Client.setMessages}.
+   *
+   * @see Client#setMessages
+   */
+  move(mailboxIds) {
+    return this._updateAllMessages({ mailboxIds });
```
Comment (dbenchi) [inline lib/models/Thread.js:85]: So I have an idea:
We have said that Thread is a virtual concept while the real one is the messages. So I propose to add updateAllMessages and flagAllMessages in a Messages === [Message] model and we can leverage them from here. Cause it is not easy for a normal user to understand why do we have such operationAllMessages in the Thread model, WDYT?

File: lib/models/Thread.js (hunk 1 above)
Comment (ddolcimascolo) [inline lib/models/Thread.js:128]: It is not an array, it's the ES6 syntax to create dynamic properties on objects.

File: lib/models/Thread.js (hunk 2 above)
Comment (ddolcimascolo) [inline lib/models/Thread.js:115]: Exactly, or even move and set flags at once. These are really the list of _updates_ to apply.

File: lib/models/Thread.js (hunk 3 above)
Comment (ddolcimascolo) [inline lib/models/Thread.js]: Not required, it is an internal method.

File: test/common/Client.js (hunk 4 above)
Comment (ddolcimascolo) [inline test/common/Client.js:2076]: No, you did not read the code :) `destroyMessages` returns the `MessageSet` directly, there's no such logic.

File: lib/models/Thread.js (hunk 5 above)
Comment (ddolcimascolo) [inline lib/models/Thread.js:85]: I don't understand.
These two methods are internal helper methods, that simply apply operations on the list of mesage ids congtained in the thread.

File: lib/models/Thread.js (hunk 3 above)
Comment (ddolcimascolo) [inline lib/models/Thread.js]: Ok for state, not for flag. This in an internal method and `flag` is passed by other methods directly, not by client code.

Comment (mmaudet) [inline lib/models/Thread.js]: Il fait si mauvais que ça du côté de poissy ? 😉

Envoyé avec Boxer couplé à OBM

Le 31 janv. 2016 11:41, ddolcimascolo notifications@github.com a écrit :

In lib/models/Thread.js:

> - destroy() { > + return this._jmap.destroyMessages(this.messageIds); > + } > + > + _updateAllMessages(updates) { > + let options = {}; > + > + this.messageIds.forEach(id => { > + options[id] = updates; > + }); > + > + return this._jmap.setMessages({ update: options }); > + } > + > + _flagAllMessages(flag, state) { > + Utils.assertRequiredParameterHasType(state, flag, 'boolean'); 

Ok for state, not for flag. This in an internal method and flag is passed by other methods directly, not by client code.

**Comment by thomas-hilaire** [comment]
:+1:

**Comment by dbenchi** [comment]
:+1:

**Pull request merged**